errors.md

```diff
@@ -0,0 +1,448 @@
+  private String databasePath;
+  private Integer shardId;
+  
+  GroomException(String databasePath, Integer shardId) {
+    this.databasePath = databasePath;
+	this.shardId = shardId;
+  }
+  
+  String getDatabasePath() {
+    return databasePath;
+  }
+  
+  Integer getShardId() {
+    return shardId;
+  }
+}
+```
+
+```go
+func (r *AccountRecords) computeMonthClose(accountId int) *Ledger {
+  var out Ledger
+  shardIds := r.databaseShards(accountId)
+  var toGroom []*GroomError
+  for shardId := range shardIds {
+    records, err := r.loadRecords(accountId, shardId)
+	switch err.(type) {
+  	case nil:
+	  out.merge(records)
+	case *GroomError:
+	  toGroom = append(toGroom, err)
+	default:
+	  r.pageOncall(err)
+	  return nil // Better to return nothing than something incorrect.
+	}
+  }
+  
+  // Retry
+  
+  for _, pending := range toGroom {
+    r.performGroom(pending.DBPath)
+	records, err := r.loadRecords(accountId, shardId)
+	if err != nil {
+	  r.pageOncall(err)
+	  return nil // Better to return nothing than something incorrect.
+	}
+	out.merge(records)
+  }
+  return &out
+}
+
+func (r *AccountRecords) loadRecords(accountId, shardId int) (*Transaction, error) {
+  dbPath := r.databasePath(accountId, shardId)
+  if (r.dataState(databasePath) != Groomed) {
+    return nil, &GroomError{DBPath: dbPath, ShardId: shardId}
+  }
+  // Rest omitted.
+}
+
+type GroomError struct{
+  DBPath string
+  ShardId int
+}
+
+func (err *GroomError) Error() string { return fmt.Sprintf("DBPath %q for ShardID %v requires grooming.", err.DBPath, err.ShardID) }
+```
+
+While the examples are contrived, they do demonstrate that there is a significant correspondence between structured error handling in Java and Go.
+Go makes authoring structured error types a breeze.  Consider:
+
+```go
+type CycleError struct {
+  Path []Node
+}
+
+func (err *CycleError) Error() string {
+  // Rest omitted.
+}
+```
+
+Not only can your debug representation of the error be nice; your APIs callers can extensibly handle defects at-will.
+
+## Reappraisal with golang.org/x/exp/errors
+
+In Go 1.13, many of the ideas in golang.org/x/exp/errors will debut.
+
+### Call Stack Diagnostic Information
+
+Out of the box up until Go 1.13, the Go standard library provide no way to perform stack trace annotation on errors.  This is a disadvantage compared to Java's Throwable with tracked stack trace, yet not
+providing such information does mean errors are treated as ordinary values, which means errors in Go traditionally are low-overhead.  From an error-handling perspective, it is rare that a program makes
+use of such stack traces for programmatic handling, aside from better informing developers and operators.
+
+Go 1.13 will change things with the introduction of golang.org/x/exp/errors.
+
+TODO: show initial call stack for root error
+TODO: show call stack for wrapped error with root error
+
+### Error-Is Matching Semantics
+
+TODO: rework sentinel matching here
```

Comment:
I would say this is actually the most important part of the new error-inspection proposal, and probably deserves more detail than the rest of the proposal (such as implementing the `Wrapper` interface for example). The idea of adding "contains(x | x == t || type(x) == t)" to the lexicon of error judgements is the real punch line of that whole spec IMO.

Reply:
Yeah, I plan on working with Damien on this point.  :)